### PR TITLE
Filters

### PR DESCRIPTION
**Pull request by MoshiKoi**
Proof of concept, looking for feedback

### CODE REVIEW

**Comment by cellio** [comment]
Addresses https://github.com/codidact/qpixel/issues/755 .

File: app/controllers/users_controller.rb [hunk 1]
```diff
@@ -64,6 +64,36 @@ def preferences
...
   end
 
+  def filters
+    system_filters = Rails.cache.fetch 'system_filters' do
+      User.find(-1).filters.to_h { |filter| [filter.name, filter.json] }
+    end
+
+    respond_to do |format|
+      format.json do
+        if current_user
```
Comment (ArtOfCode-) [inline app/controllers/users_controller.rb:74]:
```suggestion
        if user_signed_in?
```

File: app/models/filter.rb [hunk 2]
```diff
@@ -0,0 +1,15 @@
+class Filter < ApplicationRecord
+  belongs_to :user, required: true, class_name: 'User'
```
Comment (ArtOfCode-) [inline app/models/filter.rb:2]: You should be able to drop the `class_name` argument here - Rails should infer it from the `:user` name. `required` also isn't strictly necessary: any reference that isn't explicitly optional is required by default.

**Comment by MoshiKoi** [comment]
I've reached a good point to stop working on further features for this pull request, I think. I've completed most of the stuff in #755 and #756, now it's just testing (maybe spin this branch up on the dev server for people to mess about with?)

Here are the changes:

## Side widget

On the side of category feeds is the filter widget, where users can enter filter options. On submitting, the category will be filtered.

image

## Search filters

Should be fairly intuitive. Filters the results of a search. NOTE: This is *in addition* to any search queries - if the user types "score:>0.6" and filters by a minimum score of "0.5", then it will only show results with score above 0.6. The reverse applies too.

image

## System presets 

System presets are initially populated by seeding. You can take a look in https://github.com/MoshiKoi/qpixel/blob/filters/db/seeds/filters.yml for the format

Note to Administrators: There's no UI tool right now to edit the system presets after the initial population of the database. If you want to modify them, they're just normal filter records assigned to the System user (user_id: -1)

## User-created filters

If a user types in an option that isn't already the name of a preset, saving the filter will create a new filter for them. They can also modify an existing option by hitting save after changing a field.

## Management

There is now a tab on the user page where users can create/edit/delete their filters.

image

## Future Improvements

While not strictly necessary, there are some "nice to haves" I can think off for future work.

 - Don't allow users to edit System presets on the client side. (I already do validation on the backend to make sure they have permissions, but this is just a nice UX thing). Right now, it just fails silently if they try to submit an edit to a System filter.
 - Admin tool to edit System presets. I feel like they change rarely enough that doing without for now shouldn't be too much of a problem, and I already have a seed file which can populate it initially.
 - Make filters that can't apply to a category greyed out (e.g., "min answers" filter option on a category with only Articles). This is a "nice to have", but it's not really necessary. 
 - Add tag includes/excludes to filters. This is somewhat involved from a technical perspective, requiring either changes to the tags API or a new API, so I'm deferring it.
 - Add "favorite tags" option as a filter (relies on the above)

**Comment by cellio** [comment]
Thanks for all your work on this!  This is great!

This comment is based on your comment (not on seeing it live).  I would prefer that users not be able to redefine system filters for themselves; that would remove the ability to get back to the defaults.  The system filters should be immutable, _but_ users should be able to create new filters based on existing ones.  What I'm imagining is: you choose a system filer, edit some properties (maybe your "positive" includes slightly negative, like a min of 0.45), and get a "save as" option where you specify a new name.  That way you have both your version and the system version available.

This means there needs to be a place to type that new name.  I don't care if that's part of the widget or a modal that comes up when you click the button.

On a different subject, can we put the filter parameters behind an expander, so you initially just see the selector for predefined filters, but you can click the expander to see the rest?  I think most people, most of the time, will use a small number of filters and not want to see all the details all the time, which pushes other content in the sidebar down.  (If my description isn't clear enough, let me know and I'll draw a picture.)

**Comment by MoshiKoi** [comment]
> This comment is based on your comment (not on seeing it live). I would prefer that users not be able to redefine system filters for themselves; that would remove the ability to get back to the defaults.

It just silently fails if they try to redefine system filters right now.

> On a different subject, can we put the filter parameters behind an expander

No problem 👍

**Comment by MoshiKoi** [comment]
It doesn't really look as nice though....

image

**Comment by cellio** [comment]
Do you also have a screenshot of the collapsed version?

**Comment by MoshiKoi** [comment]
Her it is collapsed image

**Comment by cellio** [comment]
That looks fine to me.  You could actually make it a little more compact; "Filters" and "predefined filters" are redundant, so we could just have "filters", the selector, and the options behind the expander.  The fact that there's a list means there are some predefined ones, after all. :-)

Can we only light up "save" if there are unsaved changes?  We don't want to lure people into hitting that a bunch of times because they think they have to in order to use a filter.

**Comment by MoshiKoi** [comment]
> Can we only light up "save" if there are unsaved changes?

If I actually select something, then it greys it out - though i suppose i could grey it out on initialization as well

**Comment by MoshiKoi** [comment]
It turns out tag filters were much simpler than expected; I just misunderstood the Select2 documentation

image

**Comment by MoshiKoi** [comment]
> It turns out tag filters were much simpler than expected; I just misunderstood the Select2 documentation

It turns out my initial complexity assessment was actually not that far off... it was complicated in a different way

**Comment by cellio** [comment]
@ArtOfCode- , Moshi said in chat that this is ready for an initial code review.  I see there are merge conflicts which obviously have to be resolved, but can someone can take a look at the code for any other issues in the meantime?

File: app/controllers/users_controller.rb [hunk 3]
```diff
@@ -64,6 +64,100 @@ def preferences
...
+    if params[:system] == true
+      if current_user&.is_global_admin
+        as_user = User.find(-1)
+      else
+        return render json: { status: 'failed', success: false, errors: ['You do not have permission to delete'] },
+                      status: 400
+      end
+    end
+
+    filter = Filter.find_by(user: as_user, name: params[:name])
```
Comment (Taeir) [inline app/controllers/users_controller.rb:141]: As I understand it, it is possible to delete the filter of the system user (used as default filter for category?), but not to set it with the `set_filter` method above?

File: app/views/search/_filters.html.erb [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+                                                            selected: @active_filter[:name]),
+          include_blank: true, class: "form-element js-filter-select", id: nil,
+          data: { placeholder: "" } %>
+    </div>
+    <button type="button" class="filter-clear button is-medium is-danger">Clear</button>
+    <% if allow_apply %>
+      <%= submit_tag 'Apply', class: 'button is-medium is-outlined', name: nil %>
+    <% end %>
+    <% if user_signed_in? %>
+      <button type="button" class="filter-save button is-medium is-filled">Save</button>
```
Comment (Taeir) [inline app/views/search/_filters.html.erb:19]: Should we display the button but grayed out (disabled) and with a title set to indicate "this is not possible because you are not signed in"?

**Review by Taeir — COMMENTED** [review]
I think this is some quite great work @MoshiKoi ! There are quite a few annoying/tricky cases in the search (and other places) that you dealt with in a pretty nice way :)

I do have quite a few review comments, some of those questions about the workings, others small improvements or suggestions. For a final version, I also think we should write some tests, to verify that basic filters work as intended.

I have not yet checked your JavaScript code, I kept to the Ruby and HTML for now. I will take a look at that at a later time if I get to it.

File: app/assets/javascripts/filters.js [hunk 5]
```diff
@@ -0,0 +1,148 @@
+$(() => {
+  $('.js-filter-select').each(async (_, el) => {
+    const $select = $(el);
+    const $form = $select.closest('form');
+    const $formFilters = $form.find('.form--filter');
+    const $saveButton = $form.find('.filter-save');
+    const $isDefaultCheckbox = $form.find('.filter-is-default');
+    const categoryId = $isDefaultCheckbox.val();
+    var defaultFilter = await QPixel.defaultFilter(categoryId);
```
Comment (ArtOfCode-) [inline app/assets/javascripts/filters.js:9]: consider (es+): doesn't need to be a `var` just because it's from `await` (if that was the reason) - `const` if it's a set-once value, `let` if it will be changed later.

File: app/assets/javascripts/filters.js [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+
+        $isDefaultCheckbox.prop('checked', defaultFilter === $select.val());
+        computeEnables();
+
+        // Name is not one of the presets, i.e user is creating a new preset
+        if (!preset) {
+          return;
+        }
+
+        for (const [name, value] of Object.entries(preset)) {
```
Comment (ArtOfCode-) [inline app/assets/javascripts/filters.js:90]: consider (es+, non-essential): `Object.entries(preset).forEach((k, v) => {` etc.

File: app/assets/javascripts/qpixel_api.js [hunk 7]
```diff
@@ -263,6 +263,95 @@ window.QPixel = {
...
+      credentials: 'include',
+      headers: {
+        'Accept': 'application/json'
+      }
+    });
+    const data = await resp.json();
+    return data.name;
+  },
+
+  setFilterAsDefault: async (category_id, name) => {
```
Comment (ArtOfCode-) [inline app/assets/javascripts/qpixel_api.js:298]: request (style): there are a few variables in snake_case in this file - unless there's a reason they need to be, can we change to camelCase?

File: app/controllers/users_controller.rb [hunk 8]
```diff
@@ -64,6 +64,100 @@ def preferences
...
+        render json: filters_json
+      end
+    end
+  end
+
+  def set_filter
+    if user_signed_in? && params[:name]
+      filter = Filter.find_or_create_by(user: current_user, name: params[:name])
+
+      filter.update(min_score: params[:min_score], max_score: params[:max_score],
```
[the remark this answers is not in the document]
Comment (ArtOfCode-) [reply app/controllers/users_controller.rb:107]: No obvious security risks that I can see here, but the validation concerns are... valid (ha!). Worth considering strong params here, i.e. `filter.update(permitted_params)`, where `permitted_params` is a method that calls `params.require(...).permit(...)` etc.

File: app/controllers/users_controller.rb [hunk 9]
```diff
@@ -64,6 +64,100 @@ def preferences
...
+                    min_answers: params[:min_answers], max_answers: params[:max_answers],
+                    include_tags: params[:include_tags], exclude_tags: params[:exclude_tags],
+                    status: params[:status])
+
+      unless params[:category].nil? || params[:is_default].nil?
```
[the remark this answers is not in the document]
Comment (ArtOfCode-) [reply app/controllers/users_controller.rb:112]: This one might be more down to personal preference? I like an unless statement in Ruby - I find they make sense if you read them in natural language to yourself.

"Unless params[:category] is nil or params[:is_default] is nil"

vs

"If params[:category] is not nil and params[:is_default] is not nil"

File: app/helpers/search_helper.rb [hunk 10]
```diff
@@ -1,4 +1,85 @@
...
+  def filter_to_qualifiers(filter)
+    qualifiers = []
+    qualifiers.append({ param: :score, operator: '>=', value: filter.min_score }) unless filter.min_score.nil?
+    qualifiers.append({ param: :score, operator: '<=', value: filter.max_score }) unless filter.max_score.nil?
+    qualifiers.append({ param: :answers, operator: '>=', value: filter.min_answers }) unless filter.min_answers.nil?
+    qualifiers.append({ param: :answers, operator: '<=', value: filter.max_answers }) unless filter.max_answers.nil?
+    qualifiers.append({ param: :include_tags, tag_ids: filter.include_tags }) unless filter.include_tags.nil?
+    qualifiers.append({ param: :exclude_tags, tag_ids: filter.exclude_tags }) unless filter.exclude_tags.nil?
+    qualifiers.append({ param: :status, value: filter.status }) unless filter.status.nil?
+    qualifiers
```
Comment (ArtOfCode-) [inline app/helpers/search_helper.rb:39]: consider (quality): is it better to iterate through the properties of `filter` and add to `qualifiers` for each that exists, rather than repeated `append/unless` statements?

File: app/assets/javascripts/filters.js (hunk 5 above)
Comment (MoshiKoi) [reply app/assets/javascripts/filters.js:9]: Yes, this should be `let`; thank's for catching that

File: app/assets/javascripts/filters.js (hunk 6 above)
Comment (MoshiKoi) [reply app/assets/javascripts/filters.js:90]: My usage of `for...of` intead of `.forEach` is mostly personal preference since I think it's more readable. Is there a benefit to the `.forEach` version that I'm missing?

File: app/assets/javascripts/qpixel_api.js (hunk 7 above)
Comment (MoshiKoi) [reply app/assets/javascripts/qpixel_api.js:298]: Will do 👍

File: app/assets/javascripts/filters.js (hunk 6 above)
Comment (ArtOfCode-) [reply app/assets/javascripts/filters.js:90]: Nope, likewise that's just preference, that's why it's a consider :)

File: config/config/preferences.yml [hunk 11]
```diff
@@ -17,6 +17,7 @@ keyboard_tools:
   description: >
     Enable keyboard shortcuts. Press ? for a list of shortcuts.
   default: 'true'
+  global: true
```
[the remark this answers is not in the document]
Comment (MoshiKoi) [reply config/config/preferences.yml:23]: Done 👍

File: app/views/search/_filters.html.erb (hunk 4 above)
Comment (MoshiKoi) [reply app/views/search/_filters.html.erb:19]: Not sure - I personally prefer it the way it is now, but I'm open to changing it

File: app/helpers/search_helper.rb (hunk 10 above)
Comment (MoshiKoi) [reply app/helpers/search_helper.rb:39]: I'm not really sure how that would look to be honest; I'm not familiar enough with idiomatic Ruby patterns.

File: test/fixtures/category_filter_defaults.yml [hunk 12]
```diff
@@ -0,0 +1,11 @@
...
+
+one:
+  user: one
+  filter: one
+  category: one
+
+two:
+  user: two
+  filter: two
+  category: two
```
[the remark this answers is not in the document]
Comment (MoshiKoi) [reply test/fixtures/category_filter_defaults.yml:11]: The test fixtures were autogenerated by rails so that's why they haven't been properly set up (and also I don't know how to set up ruby tests).

File: app/views/search/_filters.html.erb [hunk 13]
```diff
@@ -0,0 +1,75 @@
+<% allow_delete ||= false %>
+<% allow_apply = true if allow_apply.nil? %>
+<% @active_filter ||= {} %>
+
+<div class="form-group">
+  <div class="form-group-horizontal">
+    <div class="form-group">
+      <%= label_tag :predefined_filter, 'Predefined Filters', class: "form-element" %>
+      <%= select_tag :predefined_filter, options_for_select([@active_filter[:name]],
```
[the remark this answers is not in the document]
Comment (MoshiKoi) [reply app/views/search/_filters.html.erb:9]: It doesn't; this is just to set the active filter as the active selection in the select element. Select2 fills in the actual options.

File: app/views/search/_filters.html.erb (hunk 4 above)
Comment (cellio) [reply app/views/search/_filters.html.erb:19]: What is the current behavior?  (She says, hoping to avoid having to pull and set it up locally to find out. :-) )

File: db/migrate/20220916075849_add_tags_to_filters.rb [hunk 14]
```diff
@@ -0,0 +1,6 @@
+class AddTagsToFilters < ActiveRecord::Migration[7.0]
+  def change
+    add_column :filters, :include_tags, :string
```
[the remark this answers is not in the document]
Comment (MoshiKoi) [reply db/migrate/20220916075849_add_tags_to_filters.rb:3]: This is the same way that tags are stored on posts and it seems to be doing okay - I don't imagine someone is going to be filtering by more tags than can fit on a post, but I could be wrong. 

https://github.com/codidact/qpixel/blob/ad595c3b2d186d77a9070894af12be050bf8f853/db/schema.rb#L411

File: app/views/search/_filters.html.erb (hunk 4 above)
Comment (MoshiKoi) [reply app/views/search/_filters.html.erb:19]: It's just Not There™️

File: test/fixtures/filters.yml [hunk 15]
```diff
@@ -0,0 +1,17 @@
+# Read about fixtures at https://api.rubyonrails.org/classes/ActiveRecord/FixtureSet.html
+
+one:
+  name: MyString
+  min_score: 1.5
```
Comment (Taeir) [inline test/fixtures/filters.yml:5]:
```suggestion
  user: standard_user
  min_score: 1.5
```

File: app/models/user.rb [hunk 16]
```diff
@@ -24,6 +24,7 @@ class User < ApplicationRecord
   has_many :comments, dependent: :nullify
   has_many :comment_threads_deleted, class_name: 'CommentThread', foreign_key: :deleted_by_id, dependent: :nullify
   has_many :comment_threads_locked, class_name: 'CommentThread', foreign_key: :locked_by_id, dependent: :nullify
+  has_many :filters, dependent: :destroy
```
Comment (Taeir) [inline app/models/user.rb:27]: Tests fail because this doesnt reference the category_filter_defaults, needs:

```suggestion
  has_many :filters, dependent: :destroy
  has_many :category_filter_defaults, dependent: :destroy
```

File: app/models/user.rb [hunk 17]
```diff
@@ -24,6 +24,7 @@ class User < ApplicationRecord
...
   belongs_to :deleted_by, required: false, class_name: 'User'
```
Comment (Taeir) [inline app/models/user.rb:28]: On another note, the question is why category_filter_defaults is linked to user, since it is also linked to filter and filter is linked to a user.

Comment (MoshiKoi) [reply app/models/user.rb:28]: The problem is the System filters. We could have, for instance

Filter(Positive, belongs to System)
CategoryFilterDefault(Moshi, Meta, Positive)

To be honest, attaching "universal" filters to System might not have been the brightest idea, since it leads to weird edge cases like this.

**Comment by MoshiKoi** [comment]
This test no longer succeeds due to [a change in how search works](https://meta.codidact.com/posts/287031):

https://github.com/codidact/qpixel/blob/8f36e90346ce1413cab8e6973620381a184c0153/test/controllers/search_controller_test.rb#L6-L10

Now, an empty search will return all posts. Should this method be changed to instead assert not nil?

**Comment by MoshiKoi** [comment]
Since it was easy, I just added in #905 as well.

File: app/helpers/search_helper.rb (hunk 10 above)
Comment (Taeir) [reply app/helpers/search_helper.rb:39]: Since the treatment of each property is different, I see no good way to do what you propose.

**Comment by MoshiKoi** [comment]
Work is continuing on #976
